### PR TITLE
Add Pascal VOC Datasets for image detection and segmentation 

### LINKED ISSUE #139: [Dataset] Add Pascal VOC Datasets for image detection and segmentation
Please add the following Pascal VOC Dataset as new _Image detection or segmentation_ datasets: 

- [ ] [Pascal VOC](http://host.robots.ox.ac.uk/pascal/VOC/) Segmentation Dataset.
- [ ] [Pascal VOC](http://host.robots.ox.ac.uk/pascal/VOC/) Detection Dataset.

### PR DESCRIPTION
**Pull request by DerrickUnleashed**
This PR adds two datasets:
`pascal_segmentation_dataset()` and `pascal_detection_dataset()`

- Selective download is implemented
- Inheritance is implemented between the datasets
- consistent according to #165 

Closes #139

### CODE REVIEW

File: R/dataset-pascal.R [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+#' @return A torch dataset of class \code{pascal_segmentation_dataset}.
+#' Each element is a named list:
+#' - `x`: a H x W x 3 array representing the RGB image.
+#' - `y`: a list with one item:
+#'     - `masks`: a torch boolean tensor of shape (3, H, W) representing the segmentation mask.
+#'
+#' @examples
+#' \dontrun{
+#' # Load Pascal VOC segmentation dataset (2012 train split)
+#' pascal_seg <- pascal_segmentation_dataset(transform = transform_to_tensor, download = TRUE)
```
Comment (cregouby) [inline R/dataset-pascal.R:20]: **improvement** could we use the parameters leading to the smallest available dataset for the eample ? (apply to both)

File: NEWS.md [hunk 2]
```diff
@@ -3,6 +3,8 @@
 ## New datasets
 
 * Added `lfw_people_dataset()` and `lfw_pairs_dataset()` for loading Labelled Faces in the Wild (LFW) datasets (@DerrickUnleashed, #203).
+* Added `places365_dataset()`for loading the Places365 dataset (@koshtiakanksha, #196).
+* Added `pascal_segmentation_dataset()`, and `pascal_detection_dataset()` for loading the Pascal Visual Object Classes datasets (@DerrickUnleashed, #139).
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply NEWS.md]: Thanks for the suggestion.

File: R/dataset-pascal.R (hunk 1 above)
Comment (DerrickUnleashed) [reply R/dataset-pascal.R:20]: Sure, I'll implement it that way. Thanks, this should improve the user experience.

File: R/dataset-pascal.R [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+      labels[i] <- xml2::xml_text(xml2::xml_find_first(obj, "name"))
+
+      bbox <- xml2::xml_find_first(obj, "bndbox")
+      xmin <- as.integer(xml2::xml_text(xml2::xml_find_first(bbox, "xmin")))
+      ymin <- as.integer(xml2::xml_text(xml2::xml_find_first(bbox, "ymin")))
+      xmax <- as.integer(xml2::xml_text(xml2::xml_find_first(bbox, "xmax")))
+      ymax <- as.integer(xml2::xml_text(xml2::xml_find_first(bbox, "ymax")))
+
+      boxes[[i]] <- c(xmin, ymin, xmax, ymax)
+    }
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-pascal.R:304]: Thanks for these improvements and suggestions. Really improved the code readability , usage and execution time.
```r
parse_voc_xml = function(xml) {
  objects <- xml2::xml_find_all(xml, ".//object")

  labels <- xml2::xml_text(xml2::xml_find_all(objects, "name"))

  bboxes <- xml2::xml_find_all(objects, "bndbox")

  xmin <- xml2::xml_integer(xml2::xml_find_all(bboxes, "xmin"))
  ymin <- xml2::xml_integer(xml2::xml_find_all(bboxes, "ymin"))
  xmax <- xml2::xml_integer(xml2::xml_find_all(bboxes, "xmax"))
  ymax <- xml2::xml_integer(xml2::xml_find_all(bboxes, "ymax"))

  boxes <- torch_tensor(data.frame(xmin, ymin, xmax, ymax) %>% as.matrix(), dtype = torch_int64())

  list(
    labels = labels,
    boxes = boxes
  )
}
```

File: R/dataset-pascal.R [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    fs::file_exists(file.path(self$processed_folder, paste0(self$split, ".rds")))
+  },
+
+  .getitem = function(index) {
+    img_path <- self$img_path[index]
+    mask_path <- self$mask_paths[index]
+
+    x <- jpeg::readJPEG(img_path)
+    masks <- png::readPNG(mask_path)*255
+    masks <- torch_tensor(masks, dtype = torch_bool())$permute(c(3, 1, 2))
```
[the remark this answers is not in the document]
Comment (cregouby) [reply R/dataset-pascal.R]: Hello @DerrickUnleashed,

**praise** The result is visually much better now ! 
**improvement** Even if strictly correct, the current output requires combining `$y$labels` and $y$masks` in order to condition 10 images in a batch for a model finetuning, which looks like a nightmare for the end-user.
**suggestion** Could we output `$y$masks` into a [ CPUBoolType{21,H, W} ] tensor, each of the 21 boolean layer beeing a label mask ? (see https://github.com/mlverse/torchvision/blob/main/R/models-fcn.R)

**todo** **performance** your code to transform the mask is taking way too long to read and transform two images into tensors. 
```
> tictoc::tic();first_item <- pascal_seg[205]; tictoc::toc()
2.235 sec elapsed
```

File: R/dataset-pascal.R [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+  .getitem = function(index) {
+    voc_colormap <- list(
+      c(0, 0, 0), c(128, 0, 0), c(0, 128, 0), c(128, 128, 0),
+      c(0, 0, 128), c(128, 0, 128), c(0, 128, 128), c(128, 128, 128),
+      c(64, 0, 0), c(192, 0, 0), c(64, 128, 0), c(192, 128, 0),
+      c(64, 0, 128), c(192, 0, 128), c(64, 128, 128), c(192, 128, 128),
+      c(0, 64, 0), c(128, 64, 0), c(0, 192, 0), c(128, 192, 0),
+      c(0, 64, 128)
+    )
```
Comment (cregouby) [inline R/dataset-pascal.R:190]: **improvement** **performance** please move this affectation out of the getitem. Please remember that getitem will be called ~16000 time for a model training task
**improvement** **performance** please do not store data into a list if you use it as an array.

File: R/dataset-pascal.R [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+      ncol = 3
+    )
+    colormap_mat <- do.call(rbind, voc_colormap)
+    match_indices <- apply(flat_mask, 1, function(px) {
+      matched <- which(colSums(abs(t(colormap_mat) - px)) == 0)
+      if (length(matched) > 0) matched[1] - 1 else NA_integer_
+    })
+    class_idx <- matrix(match_indices, nrow = dim(mask_data)[1], ncol = dim(mask_data)[2])
+    masks <- torch_tensor(class_idx, dtype = torch_long())$unsqueeze(1)
+    labels <- sort(unique(as.vector(as_array(masks)))+1)
```
Comment (cregouby) [inline R/dataset-pascal.R:210]: **improvement** **performance** **blocking** Iterating on pixel is a design antipattern. joining two dataframes with a witch on a numerical operation is a design antipattern.
**suggestion** torch is your friend for that : 
```r
mask_data <- png::readPNG(pascal_seg$mask_paths[17]) *255
# on  a small portion of a mask image with multiple colors
 mask_t <- mask_data[180:185, 113:123, ] |> torch_tensor(dtype = torch_long()) |> torch_movedim(c(3,1,2), c(1,2,3))
# torch can do the color binary matching (one color in this example)
(mask_t == torch_ones_like(mask_t)$mul(torch_tensor(c(192,0,128))$unsqueeze(2)$unsqueeze(3)))$sum(dim=1) == 3
```

File: R/dataset-pascal.R [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    masks <- masks$to(dtype = torch_bool())
```
Comment (cregouby) [inline R/dataset-pascal.R:211]: **suggestion** this will be useless as soon as masks is the result of a torch boolean operation.

**Review by cregouby — CHANGES_REQUESTED** [review]
**todo** major performance improvement is needed. Some hints are included.

File: R/dataset-pascal.R (hunk 5 above)
Comment (DerrickUnleashed) [reply R/dataset-pascal.R:190]: Thanks for the improvement. This has made the code way more efficient also.
```r
> tic();first_item <- ds[2];toc();
0.016 sec elapsed
```

File: R/dataset-pascal.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-pascal.R]: Thanks for the improvement. I agree it would be better if the Masks have {21,H,W}.
I have made the following updates.
```r
> tic(); item <- ds[174] ; toc();
0.043 sec elapsed
```
```r
> first_item$y$masks$shape
[1]  21 375 500
```
```r
ds <- pascal_segmentation_dataset(
  year = "2007",
  split = "test",
  download = TRUE,
  transform = transform_to_tensor
)
first_item <- ds[200]
masked_img <- draw_segmentation_masks(first_item,colors = c(
       "red", "green", "blue", "yellow", "magenta", "cyan", "orange", "purple",
       "brown", "pink", "grey", "navyblue", "maroon", "turquoise", "limegreen",
       "gold", "darkgreen", "skyblue", "orchid", "black", "deepskyblue"
   ))
tensor_image_browse(masked_img)
```
image

File: R/dataset-pascal.R (hunk 7 above)
Comment (DerrickUnleashed) [reply R/dataset-pascal.R:211]:
```r
match_indices <- match(rgb_to_int(flat_mask), rgb_to_int(colormap_mat)) - 1
class_idx <- matrix(match_indices, nrow = dim(mask_data)[1], ncol = dim(mask_data)[2])
class_idx_tensor <- torch_tensor(class_idx, dtype = torch_long())
masks <- torch_zeros(21, dim(class_idx_tensor)[1], dim(class_idx_tensor)[2], dtype = torch_bool())
for (i in 0:20) {
  masks[i + 1][class_idx_tensor == i] <- TRUE
}
labels <- which(as_array(masks$any(dim = c(2, 3))))
```
Is this fine or do I need to make some improvements ?

File: R/dataset-pascal.R [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+#' @inheritParams oxfordiiitpet_dataset
+#' @param root Character. Root directory where the dataset will be stored under `root/pascal_voc_<year>`.
+#' @param year Character. VOC dataset version to use. One of `"2007"`, `"2008"`, `"2009"`, `"2010"`, `"2011"`, or `"2012"`. Default is `"2012"`.
+#' @param split Character. One of `"train"`, `"val"`, `"trainval"`, or `"test"`. Determines the dataset split. Default is `"train"`.
+#'
+#' @return A torch dataset of class \code{pascal_segmentation_dataset}.
+#' Each element is a named list:
+#' - `x`: a H x W x 3 array representing the RGB image.
+#' - `y`: a list with one item:
+#'     - `masks`: a torch boolean tensor of shape (3, H, W) representing the segmentation mask.
```
Comment (cregouby) [inline R/dataset-pascal.R:15]: **todo** please reflect the new `$masks` format and shape and encoding in the function documentation?

File: R/dataset-pascal.R [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+#' The Pascal Visual Object Classes (VOC) dataset is a benchmark in object recognition and segmentation tasks.
+#' This dataset provides images with per-pixel class segmentation masks for 20 object categories across different years (2007–2012).
+#'
+#' @inheritParams oxfordiiitpet_dataset
+#' @param root Character. Root directory where the dataset will be stored under `root/pascal_voc_<year>`.
+#' @param year Character. VOC dataset version to use. One of `"2007"`, `"2008"`, `"2009"`, `"2010"`, `"2011"`, or `"2012"`. Default is `"2012"`.
+#' @param split Character. One of `"train"`, `"val"`, `"trainval"`, or `"test"`. Determines the dataset split. Default is `"train"`.
+#'
+#' @return A torch dataset of class \code{pascal_segmentation_dataset}.
+#' Each element is a named list:
```
Comment (cregouby) [inline R/dataset-pascal.R:12]: **todo** **missing** please add a mention to the class of each item and thus its ability to be a direct input to `draw_segmentation_masks`

File: R/dataset-pascal.R [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+#'
+#' The Pascal Visual Object Classes (VOC) dataset is a benchmark in object recognition and segmentation tasks.
+#' This dataset provides images with per-pixel class segmentation masks for 20 object categories across different years (2007–2012).
+#'
+#' @inheritParams oxfordiiitpet_dataset
+#' @param root Character. Root directory where the dataset will be stored under `root/pascal_voc_<year>`.
+#' @param year Character. VOC dataset version to use. One of `"2007"`, `"2008"`, `"2009"`, `"2010"`, `"2011"`, or `"2012"`. Default is `"2012"`.
+#' @param split Character. One of `"train"`, `"val"`, `"trainval"`, or `"test"`. Determines the dataset split. Default is `"train"`.
+#'
+#' @return A torch dataset of class \code{pascal_segmentation_dataset}.
```
Comment (cregouby) [inline R/dataset-pascal.R:11]: **todo** **missing** please also mention that the class names are present in the object.

File: R/dataset-pascal.R [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+    `2012` = list(
+      trainval = list(url = "https://huggingface.co/datasets/JimmyUnleashed/Pascal_VOC/resolve/main/VOCtrainval_11-May-2012.tar",md5 = "6cd6e144f989b92b3379bac3b3de84fd")
+    )
+  ),
+  classes = c(
+    "background", "aeroplane", "bicycle", "bird", "boat", "bottle",
+    "bus", "car", "cat", "chair", "cow", "diningtable", "dog",
+    "horse", "motorbike", "person", "pottedplant", "sheep",
+    "sofa", "train", "tvmonitor"
+  ),
```
Comment (cregouby) [inline R/dataset-pascal.R:89]: **todo** could you switch to `classes=voc_segmentation_classes` as currently there are non consistent with yours : 
```
> waldo::compare(pascal_det$classes, torchvision:::voc_segmentation_classes)
`old[9:21]`: "cat" "chair" "cow" "diningtable"  "dog" "horse" "motorbike" "person" "pottedplant"  "sheep" and 3 more...
`new[9:21]`: "cat" "chair" "cow" "dining table" "dog" "horse" "motorbike" "person" "potted plant" "sheep" ...          

```

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** This is much better now. Thanks for the update
**todo** see inline

File: R/dataset-pascal.R (hunk 11 above)
Comment (DerrickUnleashed) [reply R/dataset-pascal.R:89]:
```r
  classes = c(
    "background", "aeroplane", "bicycle", "bird", "boat", "bottle",
    "bus", "car", "cat", "chair", "cow", "dining table", "dog",
    "horse", "motorbike", "person", "potted plant", "sheep",
    "sofa", "train", "tv"
  ),
  ```

Comment (DerrickUnleashed) [reply R/dataset-pascal.R:89]: It would help if I know where the voc_segmentation_classes are from so I can make them consistent.

File: R/dataset-pascal.R (hunk 8 above)
Comment (DerrickUnleashed) [reply R/dataset-pascal.R:15]: Sure I should have updated this thank you for reminding.
